pkg/gui/presentation/commits.go

```diff
@@ -78,24 +79,76 @@ func GetCommitListDisplayStrings(
 	// function expects to be passed the index of the commit in terms of the `commits` slice
 	var getGraphLine func(int) string
 	if showGraph {
-		// this is where the graph begins (may be beyond the TODO commits depending on startIdx,
-		// but we'll never include TODO commits as part of the graph because it'll be messy)
-		graphOffset := max(startIdx, rebaseOffset)
-
-		pipeSets := loadPipesets(commits[rebaseOffset:])
-		pipeSetOffset := max(startIdx-rebaseOffset, 0)
-		graphPipeSets := pipeSets[pipeSetOffset:max(endIdx-rebaseOffset, 0)]
-		graphCommits := commits[graphOffset:endIdx]
-		graphLines := graph.RenderAux(
-			graphPipeSets,
-			graphCommits,
-			selectedCommitHash,
-		)
-		getGraphLine = func(idx int) string {
-			if idx >= graphOffset {
-				return graphLines[idx-graphOffset]
-			} else {
-				return ""
+		if len(commits) > 0 && commits[0].Divergence != models.DivergenceNone {
+			// Showing a divergence log; we know we don't have any rebasing
+			// commits in this case. But we need to render separate graphs for
+			// the Local and Remote sections.
+			allGraphLines := []string{}
+
+			_, localSectionStart, found := lo.FindIndexOf(
+				commits, func(c *models.Commit) bool { return c.Divergence == models.DivergenceLeft })
+			if !found {
+				localSectionStart = len(commits)
+			}
+
+			if localSectionStart > 0 {
+				// we have remote commits at all
```

Comment:
Happy to change it (see 440022a5c85d), but I'd be curious what you think was wrong with my version. Maybe there's something I don't understand about how English works? (What I meant to say was "this is the section of code that deals with the remote commits, so let's first check if we have any at all.")

Reply:
So I hadn't actually thought very hard about it until just now, but reading up here: https://dictionary.cambridge.org/grammar/british-grammar/at-all

> At all means ‘in any way’. We use it with questions and negatives to add emphasis, but not with affirmative statements

I think 'let's check if we have any at all' sounds fine because it implies the question 'are there any at all?' but if you have a statement 'yes, we have any at all' it feels wrong.

My understanding is that you've said 'at all' to add emphasis because it would be especially silly for us to execute all this code that depends on there being remote commits if there are no remote commits. Seems like an arbitrary shortcoming of English: I can imagine an alternative history where you could use 'at all' in affirmative statements and there wouldn't be any asymmetry but it seems like for some strange reason that's the convention.

Reply:
Good to know, thanks for the explanation.